src/puya/awst_build/eb/box/_util.py

```diff
@@ -0,0 +1,77 @@
+from puya.awst import wtypes
+from puya.awst.nodes import (
+    BoxKeyExpression,
+    BoxLength,
+    IntrinsicCall,
+    Literal,
+    SingleEvaluation,
+    UInt64Constant,
+)
+from puya.awst_build.eb.base import (
+    BuilderBinaryOp,
+    ExpressionBuilder,
+)
+from puya.awst_build.eb.var_factory import var_expression
+from puya.awst_build.utils import (
+    eval_slice_component,
+)
+from puya.errors import CodeError
+from puya.parse import SourceLocation
+
+
+def index_box_bytes(
+    box_key: BoxKeyExpression,
+    index: ExpressionBuilder | Literal,
+    location: SourceLocation,
+) -> ExpressionBuilder:
+    len_expr = BoxLength(box_key=box_key, source_location=location)
+
+    begin_index_expr = eval_slice_component(len_expr, index, location)
+    assert begin_index_expr, "Index expression cannot evaluate to None"
+    return var_expression(
+        IntrinsicCall(
+            op_code="box_extract",
+            stack_args=[
+                box_key,
+                begin_index_expr,
+                UInt64Constant(value=1, source_location=location),
+            ],
+            source_location=location,
+            wtype=wtypes.bytes_wtype,
+        )
+    )
+
+
+def slice_box_bytes(
```

Comment:
This produces some ugly awst, but I moved away from using SliceExpression as it it was expecting `base` to be a `bytes` type which it isn't always (at last not directly). I wonder if it's worth adding `BoxSlice` as a node to clean this up.  🤔

Reply:
The supported types in the AWST validation are purely for early validation, if it makes sense to add a new supported type here, we can use SliceExpression.

That said, I need some more time to review this PR. I think some of the abstractions might be leaking a bit, and especially with the planned TypeScript work starting shortly, it's important to get them right. So that might change what we do here too.